### PR TITLE
Add a query_access handler to limit lists

### PR DESCRIPTION
**Pull request by dawehner**
* [x] Ready for review
* [ ] Ready for merge

The access control handler of entities enable to check access based upon some conditions.

These access control handlers though are not applied to list queries produced by entity query and views. The node entity type provides one solution for this (node grant access), but its overly complex for most usecases.

Commerce in D7 had good experience with basically filtering by properties.

Architecture:

* Add a ```Condition``` object which describes a list of conditions (field, value, operator) together with a conjunction (AND/OR) and nested conditions.
* Add a ```query_access``` handler type which return a ```Condition``` object. Entity type should try to implement one which matches their access handlers
* Add some glue code to alter the sql queries produce by entity query and views. This glue code reaches out to the query access handler to get the conditions it should apply.

Steps in the patch:
* [x] Add a query access handler for the cacheable entity permissions
* [x] Add a query access handler for the uncacheable entity permissions

Problems:
* [ ] How do you actually add cacheability metadata to entity queries ?!

Potential todos:
  * [ ]  Expand the test coverage starting with initial OR conditions on the entity query itself

### CODE REVIEW

File: entity.module [hunk 1]
```diff
@@ -1,6 +1,54 @@
...
  */
+
+/**
+ * Implements hook_query_TAG_alter().
+ *
+ * Enables to take into account query_access handlers for entity queries.
+ */
+function entity_query_entity_query_alter(Select $query) {
+  $entity_type_id = $query->getMetaData('entity_type');
+  if (!$query->hasTag($entity_type_id . '_access')) {
```
Comment (bojanz) [inline entity.module]: Do we also need to care about the generic entity_access tag?

File: entity.module [hunk 2]
```diff
@@ -1,6 +1,54 @@
...
+  $entity_type = $entity_type_manager->getDefinition($entity_type_id);
+
+  SqlQueryAlter::create(\Drupal::getContainer())->queryAlter($query, $entity_type);
+}
+
+/**
+ * Implements hook_query_TAG_alter().
+ */
+function entity_query_views_alter(Select $select) {
+  if ($entity_type = _entity_get_entity_type_id_from_table(reset($select->getTables())['table'])) {
```
Comment (bojanz) [inline entity.module]: This line is doing way too many things at once :)

File: src/Query/PerBundleQueryAccessHandler.php [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    // When we would write down \Drupal\entity\EntityAccessControlHandler::checkEntityOwnerPermissions
+    // as boolean logic we would end up with something like this:
+    // (view any ... = 1)
+    //   || (*.uid = account.uid && operation any ... = 1)
+    //   || (*.bundle = bundle1 and (operation any bundle1 == 1))
+    //   || (*.bundle = bundle1 and *.uid = account.uid and (operation any bundle1 == 1))
+    //   || (*.bundle = bundle2 and (operation any bundle2 == 1))
+    
+    // We can though take a couple of shortcuts by checking things in PHP which don't vary per
```
Comment (bojanz) [inline src/Query/PerBundleQueryAccessHandler.php]: We can remove this shortcut comment, it feels obvious.

File: src/Query/PerBundleQueryAccessHandler.php [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    // entity.
+    
+    // No conditions are needed when the user can access all entities anyway.
+    $entity_type_id = $this->entityType->id();
+     if ($account->hasPermission("$operation any {$entity_type_id}")) {
```
Comment (bojanz) [inline src/Query/PerBundleQueryAccessHandler.php]: We also need to check the admin permission, right?

File: src/Query/PerBundleQueryAccessHandler.php [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    $bundle_info = $this->bundleInfo->getBundleInfo($entity_type_id);
+    $bundle_key = $this->entityType->getKey('bundle');
+    $has_conditions = FALSE;
+    if ($this->entityType->entityClassImplements(EntityOwnerInterface::class)) {
+      $uid_key = $this->entityType->getKey('uid');
+
+
+      // View own $entity_type permission.
```
Comment (bojanz) [inline src/Query/PerBundleQueryAccessHandler.php]: This class seems to be confused about whether it's just PerBundle, or handling both Per bundle, and EntityType granularities. This is the same dilemma we have in the core permission provider issue. I am not convinced of the need to have two classes, but we could at least have two methods, like we have right now in the module permission provider

File: entity.module (hunk 1 above)
Comment (dawehner) [reply entity.module]: I don't think so:
``` 
    if ($this->accessCheck) {
      $this->sqlQuery->addTag($this->entityTypeId . '_access');
    }
```

File: entity.module (hunk 2 above)
Comment (dawehner) [reply entity.module]: Yeah, this is quite haskell like :) Good point

File: src/Query/PerBundleQueryAccessHandler.php (hunk 3 above)
Comment (dawehner) [reply src/Query/PerBundleQueryAccessHandler.php]: Fair. I think the logic above is still useful, do you agree?

**Comment by bojanz** [comment]
This PR now needs a rebase.

Opened an issue on drupal.org: https://www.drupal.org/node/2909970

**Comment by bojanz** [comment]
Continued in #52.

**Pull request closed without merging**